### PR TITLE
Add --lint-level cli option

### LINKED ISSUE #359: Add a command-line flag to override the major/minor/patch release detection logic
> Here's an idea: Add a command-line flag to say "assume that the next release is a minor release". Then we can just pass that in CI rather than messing around with the `Cargo.toml`.
>
> We don't need the patch-release checks, and if we ever want to make a major release of anything again, we can turn off the semver check for that crate temporarily.

_Originally posted by @Darksonn in https://github.com/tokio-rs/tokio/issues/5437#issuecomment-1424325701_

Possible workaround for users until we can come up with a more full-fledged solution for https://github.com/obi1kenobi/cargo-semver-checks/issues/58

### PR DESCRIPTION
**Pull request by Finomnis**
Closes #359

### CODE REVIEW

**Comment by obi1kenobi** [comment]
Seems reasonable to me. Deferring to @epage on the final decision since this is somewhat relevant to the merge into cargo, and he hasn't had a chance to chime in on this yet.

In a chat with @tonowak, we realized that an integration test over any one of the test crates for testing `#[must_use]` (e.g. [this one](https://github.com/obi1kenobi/cargo-semver-checks/tree/main/test_crates/struct_must_use_added)) can be used to test this feature: running with `--lint-level major` should not report any lints across the `old/new` version of that crate.

If you'd like to add that test, great! If not, I can add it post-merge.

**Comment by Finomnis** [comment]
@obi1kenobi I just tried and failed ... If you could add it, that would be great. Feel free to push directly to this PR.

File: src/check_release.rs [hunk 1]
```diff
@@ -113,6 +114,7 @@ pub(super) fn run_check_release(
     let queries_to_run: Vec<_> = queries
         .iter()
         .filter(|(_, query)| !version_change.supports_requirement(query.required_update))
+        .filter(|(_, query)| query.relevant_at_lint_level(lint_level))
```
Comment (epage) [inline src/check_release.rs:117]: Don't these two checks to the same thing, just from different sources?  I feel like the lint_level flag should be `Option<LintLevel>` and if its `None`, then in the above code we use the `version_change` logic to set it. We then only need one `filter` on the query to perform.

File: src/main.rs [hunk 2]
```diff
@@ -277,6 +284,20 @@ enum SemverChecksCommands {
     CheckRelease(CheckRelease),
 }
 
+#[derive(Copy, Clone, ValueEnum, Debug)]
+enum LintLevel {
+    Major,
+    Minor,
+}
+impl std::fmt::Display for LintLevel {
```
Comment (epage) [inline src/main.rs:292]: FYI you can use `ValueEnum` to implement this: https://github.com/clap-rs/clap/blob/master/src/util/color.rs#L67

File: src/main.rs [hunk 3]
```diff
@@ -277,6 +284,20 @@ enum SemverChecksCommands {
     CheckRelease(CheckRelease),
 }
 
+#[derive(Copy, Clone, ValueEnum, Debug)]
+enum LintLevel {
+    Major,
+    Minor,
```
Comment (epage) [inline src/main.rs:290]: imo we should have a `Patch` level, even if we don't use it **yet** but I leave that to you all

File: src/check_release.rs (hunk 1 above)
Comment (Finomnis) [reply src/check_release.rs:117]: I thought this was the whole point. We don't want to override it, only filter it. If the version changed by a major increment, we don't want to run minor checks no matter what the lint level is set to.

Comment (Finomnis) [reply src/check_release.rs:117]: You are proposing pretty much exactly the change I had earlier (`--assume-semver`) which was rejected for that reason.

File: src/main.rs (hunk 3 above)
Comment (Finomnis) [reply src/main.rs:290]: That doesn't make sense, what would a `Patch` level lint be? A `Minor` lit checks if the API changes require a minor version increment. There is nothing less than a `Patch` level increment; so checking for things that would require a `Patch` level increment. That's why `RequiredSemverUpdate` only has those two options as well.

I think there still is a misunderstanding / misconsensus what we are trying to achieve here. I think we are talking on different pages.

The essential question for me is: do you agree with [this statement](https://github.com/obi1kenobi/cargo-semver-checks/issues/359#issuecomment-1424966450)?

Comment (Finomnis) [reply src/main.rs:290]: @epage As I said earlier, there are two things that seem to cause confusion in the discussions:
- Overriding the version difference vs additional filtering after the version check
- The confusing correspondence between filter levels, lint levels and version increment levels.
  - `Minor` lints check if a `minor` version increment is required. Those checks should be silenced if a `minor` change was already detected
    - `--lint-level major` means that only `major` tests should run, meaning we only check whether or not we need a `major` version increment. Therefore this flag the same effect as a `minor` version increment (this is the part that confuses people, I think; but without looking at the implementation details, it's the most intuitive way)
  - `Major` lints check if a `major` version increment is required; Those checks should be silenced if a `major` change was detected

Maybe this is more intuitive:
   - `--lint-level minor` checks for both `major` and `minor` lints and fails if the new version in `Cargo.toml` does not match those changes
   - `--lint-level major` only checks for `major` lints and ignores `minor` lints. It also only fails if the version in `Cargo.toml` does not match those changes.

Comment (Finomnis) [reply src/main.rs:290]: Alternatively, we could introduce a flag `--skip-minor-lints` that would do the same thing as `--lint-level major`. Maybe that would be more intuitive.

File: src/check_release.rs (hunk 1 above)
Comment (epage) [reply src/check_release.rs:117]: This is being further discussed in the issue

**Comment by Finomnis** [comment]
Rejected in favor of #361

**Pull request closed without merging**